mono/utils/mono-threads-posix.c

```diff
@@ -300,6 +309,36 @@ mono_native_thread_join (MonoNativeThreadId tid)
 	return !pthread_join (tid, &res);
 }
 
+void
+mono_memory_barrier_process_wide (void)
+{
+	int status;
+
+	status = pthread_mutex_lock (&memory_barrier_process_wide_mutex);
+	g_assert (status == 0);
+
+	if (memory_barrier_process_wide_helper_page == NULL) {
+		status = posix_memalign(&memory_barrier_process_wide_helper_page, PAGESIZE, PAGESIZE);
+		g_assert (status == 0);
+	}
+
+	// Changing a helper memory page protection from read / write to no access
+	// causes the OS to issue IPI to flush TLBs on all processors. This also
+	// results in flushing the processor buffers.
+	status = mprotect (memory_barrier_process_wide_helper_page, PAGESIZE, PROT_READ | PROT_WRITE);
+	g_assert (status == 0);
+
+	// Ensure that the page is dirty before we change the protection so that
+	// we prevent the OS from skipping the global TLB flush.
+	__sync_add_and_fetch ((size_t*)memory_barrier_process_wide_helper_page, 1);
```

Comment:
This doesn't have to be `sync`.
I suggest `*(volatile size_t*)p += 1`;
But I guess `sync` is ok.

Reply:
You are right. Since the whole code is guarded by mutex a non-atomic operation would work here. However, all other implementations of this technique (CoreRT, CoreCLR, membarrier-rs) use some form of atomic primitive so I prefer to keep it consistent.